ssl-hep/ServiceX_App
# Secure REST Endpoints

Partial solution to [Issue 61](https://github.com/ssl-hep/ServiceX/issues/61)

# Problem
In order to open up access to ServiceX without Kubernetes Port-Forwarding, we have to have a way to authenticate users. 

# Approach
Added a user table to the model. As a simple starting point, there is an unsecured endpoint for registering a new user and their password. The password is encrypted with SHA-256

Users can then log-in with the `/login` endpoint and receive an access token and a refresh token.

Split the Flask routes into public and internal endpoints. The public endpoints are now protected with a check for a valid JWT token.

A new config option `ENABLE_AUTH` determines if the token is required. If set to False, then all requests are accepted.

Assume that the ingress will prevent outside users from accessing the internal endpoints.

# How to Test
1. Deploy the app with `ENABLE_AUTH` set to True
2. Submit a new user registration with a POST to `/registration` and a JSON body of
```
{
	"username": "bengal",
	"password": "foobar"
}
```
3. You should see a response with a token and refresh token
4. Try logging in with the same credentials via a POST to the `/login` endpoint
5. Submit a transform request and note that you are rejected with missing credentials
6. Add a header to your transform request 
```
Authorization: Bearer <the token you got back from the login>
```
